clap_builder/src/output/help_template.rs

```diff
@@ -892,44 +890,95 @@ impl HelpTemplate<'_, '_> {
                 subcommand,
             );
         }
-        for (_, subcommand) in ord_v {
-            if !*first {
-                self.writer.push_str("\n\n");
-            }
-            *first = false;
 
-            let heading = subcommand.get_usage_name_fallback();
-            let about = subcommand
-                .get_about()
-                .or_else(|| subcommand.get_long_about())
-                .unwrap_or_default();
+        let custom_sc_headings = self.cmd.get_subcommand_custom_help_headings();
 
-            let _ = write!(self.writer, "{header}{heading}:{header:#}",);
-            if !about.is_empty() {
-                let _ = write!(self.writer, "\n{about}",);
-            }
+        // Commands under default heading
+        self.write_flat_subcommands_under_heading(&ord_v, None, first, false);
 
-            let args = subcommand
-                .get_arguments()
-                .filter(|arg| should_show_arg(self.use_long, arg) && !arg.is_global_set())
-                .collect::<Vec<_>>();
-            if !args.is_empty() {
-                self.writer.push_str("\n");
-            }
+        // Commands under custom headings
+        for heading in custom_sc_headings {
+            self.write_flat_subcommands_under_heading(&ord_v, Some(heading), first, false);
+        }
 
-            let mut sub_help = HelpTemplate {
-                writer: self.writer,
-                cmd: subcommand,
-                styles: self.styles,
-                usage: self.usage,
-                next_line_help: self.next_line_help,
-                term_w: self.term_w,
-                use_long: self.use_long,
-            };
-            sub_help.write_args(&args, heading, option_sort_key);
-            if subcommand.is_flatten_help_set() {
-                sub_help.write_flat_subcommands(subcommand, first);
-            }
+        // Help command
+        self.write_flat_subcommands_under_heading(&ord_v, None, first, true);
+    }
+
+    fn write_flat_subcommands_under_heading(
```

Comment:
If you are extrating logic to functions, please do that in a separate commit before this one

I'm also generally not a fan of functions inside of functions